### DISCUSSION TITLE
Nuget management over transitive dependency versions

### DISCUSSION THREAD

**Discussion by Jrjuniorjr** [NuGet/Home#12106] [Q&A]
I have a question about Nuget and its management.

Let's assume two scenarios here:

I have two Nuget packages A and B in my application. They depend on C which is also a Nuget package, but I don't need to declare the C package in my application.

Scenario 1. A depends on C in version 1.0. B depends on C in version 2.0. How will Nuget handle this situation?

Scenario 2. A depends on C in version 1.0. B depends on C in version 1.0. Can I upgrade package C which is used in package B and for A and rest of the project still version 1.0?

Thanks a lot for all the answers!

**Comment by zivkan** [comment]
Both questions are answered in our docs for dependency resolution: https://learn.microsoft.com/nuget/concepts/dependency-resolution

1. See the "lowest applicable version" section
2. See the "direct dependency wins" section

**Reply by Jrjuniorjr** [reply]
I better understand how Nuget works. Thanks so much for the answer @zivkan! It helped a lot!

But I have two ideas and I wanted to know what you think of it.

In scenario 1, let's assume that package A has dependency on C and has to be version 1.0 and only 1.0. B is dependent on C and can only be version 2.0.

The idea: when my app uses package A, will Nuget use package C version 1.0 and when my app uses package B will it use package C version 2.0?

What do you think of this idea?

And then another idea, for scenario 2, would be if the developer wants to update the version of just the C package that is used in the package he can do:

` <PackageReference Include="A" Version="6.0.9" TransitiveDependencies="["C": 3.0, "D": 4.0]" />
`

If package C of package A cannot be updated because it only works on version C 1.0, it should get an error. And it can alert the developer if something is possible to break.

I think this can help developers avoid version conflicts, and also makes Nuget more customized for developers, but what do you think of these two ideas?

Thank you very much!

**Reply by Jrjuniorjr** [reply]
Is there a way to do that?
Also, another question: I think there's something to help us with this situation in Java and they called shadow dependencies. Is there something like that?

**Reply by erdembayar** [reply]
There is a prototype code for shading: https://github.com/jeffkl/packageshadingprototype

Fyi @jeffkl
